kratos/containers/pointer_vector_set.h

```diff
@@ -558,54 +558,153 @@ class PointerVectorSet final
+    {
+        if (empty()) {
+            // the dataset is empty. So use push back.
+            mData.push_back(value);
+            mSortedPartSize = mData.size();
+            return iterator(mData.end() - 1);
+        } else if (position_hint == cend()) {
+            // trying to insert at the end.
+            if (KeyOf(*(position_hint - 1)) < KeyOf(*value)) {
+                // key at the position hint is less than the value of key. Hence position hint
+                // is valid. So using the push back.
+                mData.push_back(value);
+                mSortedPartSize = mData.size();
                 return iterator(mData.end() - 1);
+            } else {
+                // given position is invalid. Hence, discarding the hint.
+                return insert(value);
             }
-
-        *i = pData;
-        return i;
+        } else if (EqualKeyTo(KeyOf(*position_hint))(*ptr_begin())) {
+            // trying to insert at the front.
+            if (KeyOf(*value) < KeyOf(*position_hint)) {
+                // key at the position hint is greater than the value of key. Hence position hint
+                // is valid. So using the push back.
+                mSortedPartSize = mData.size() + 1;
+                return mData.insert(mData.begin(), value);
+            } else {
+                // given position is invalid. Hence, discarding the hint.
+                return insert(value);
+            }
+        } else {
+            // trying to insert at an arbitrary position.
+            if (KeyOf(*value) < KeyOf(*position_hint) && (KeyOf(*(position_hint - 1)) < KeyOf(*value))) {
+                mSortedPartSize = mData.size() + 1;
+                return mData.insert(mData.begin() + (position_hint - cbegin()), value);
+            } else {
+                // given position is invalid. Hence, discarding the hint.
+                return insert(value);
+            }
+        }
     }
 
     /**
      * @brief Insert elements from a range of iterators.
-     * @details This function inserts elements from a range defined by the iterators `First` and `Last`
-     * into the set. It uses the `insert` function to insert each element.
-     * @param First An input iterator pointing to the beginning of the range to insert.
-     * @param Last An input iterator pointing to the end of the range to insert.
+     * @details This function inserts element pointers from a range defined by the iterators `first` and `last`
+     * into the set. This will not insert any elements in the range, if there exists an element with a key
+     * which is equal to an element's key in the input range.
+     * @param first An input iterator pointing to the beginning of the range to insert.
+     * @param last An input iterator pointing to the end of the range to insert.
      */
     template <class InputIterator>
-    void insert(InputIterator First, InputIterator Last)
+    void insert(InputIterator first, InputIterator last)
```

Comment:
Shall we have an additional argument (or a separate method) for `insert_sorted_values`?

Reply:
It is little bit dangerous. Thats why I put `insert(const PointVectorSet& rOther)` overload to avoid doing sorting and making it unique. `insert(end(), value)` is having `O(N)` cost if we are inserting a sorted array which can be used to construct a quick sorted `PointerVectorSet`.

Reply:
Agree with @sunethwarna that adding ordered set of entities is far faster than insert without assuming that. However creating the first PointerVectorSet would involve the sort.

Disagree with @matekelemen in the priority of robustness over performance for this case that the complexity of the algorithm changes from N to NlogN. Please note that in most of the mesh generation modelers we are adding millions of entities (to sub-modelparts) which we create with ordered indices (starting from the largest) and the first sort of those is an unnecessary operation

Nevertheless, I agree that this can be added afterward